From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] mciqtz32: Seeking, set, status, where, step, cue and info commands.

The DirectShow MCI driver seeks to the end and to MCI_TO, sets door,
audio (on/off, volume, balance), video and speed, reports tracks, ready,
current track and media present, the source/destination maxima, steps,
cues, answers info, realize and setvideo, and returns
MCIERR_UNSUPPORTED_FUNCTION for the commands it does not support.
---
 dlls/mciqtz32/mciqtz.c         | 318 +++++++++++++++++++++++++++++++++--------
 dlls/mciqtz32/mciqtz_private.h |   4 +
 2 files changed, 265 insertions(+), 57 deletions(-)

diff --git a/dlls/mciqtz32/mciqtz.c b/dlls/mciqtz32/mciqtz.c
index 6fb9c2f..ef4f873 100644
--- a/dlls/mciqtz32/mciqtz.c
+++ b/dlls/mciqtz32/mciqtz.c
@@ -137,6 +137,7 @@ static DWORD MCIQTZ_drvClose(DWORD dwDevID)
         mciFreeCommandResource(wma->command_table);
         mciSetDriverData(dwDevID, 0);
         CloseHandle(wma->stop_event);
+        free(wma->file);
         HeapFree(GetProcessHeap(), 0, wma);
         return 1;
     }
@@ -326,6 +327,9 @@ static DWORD MCIQTZ_mciOpen(UINT wDevID, DWORD dwFlags,
 
     if (!create_window(wma, dwFlags, lpOpenParms))
         goto err;
+    free(wma->file);
+    wma->file = wcsdup(lpOpenParms->lpstrElementName);
+    wma->audio_off[0] = wma->audio_off[1] = FALSE;
     wma->opened = TRUE;
 
     if (dwFlags & MCI_NOTIFY)
@@ -560,14 +564,26 @@ static DWORD MCIQTZ_mciSeek(UINT wDevID, DWORD dwFlags, LPMCI_SEEK_PARMS lpParms
 
     MCIQTZ_mciStop(wDevID, MCI_WAIT, NULL);
 
+    /* positions are set in reference time */
+    IMediaSeeking_SetTimeFormat(wma->seek, &TIME_FORMAT_MEDIA_TIME);
+
     if (dwFlags & MCI_SEEK_TO_START) {
         newpos = 0;
     } else if (dwFlags & MCI_SEEK_TO_END) {
-        FIXME("MCI_SEEK_TO_END not implemented yet\n");
-        return MCIERR_INTERNAL;
+        if (FAILED(hr = IMediaSeeking_GetDuration(wma->seek, &newpos))) {
+            WARN("Cannot get the duration (hr = %lx)\n", hr);
+            return MCIERR_INTERNAL;
+        }
     } else if (dwFlags & MCI_TO) {
-        FIXME("MCI_TO not implemented yet\n");
-        return MCIERR_INTERNAL;
+        if (wma->time_format == MCI_FORMAT_FRAMES) {
+            REFTIME frame = 0.0;
+
+            if (FAILED(IBasicVideo_get_AvgTimePerFrame(wma->vidbasic, &frame)) || frame <= 0.0)
+                return MCIERR_UNSUPPORTED_FUNCTION;
+            newpos = (LONGLONG)(lpParms->dwTo * frame * 10000000.0 + 0.5);
+        } else {
+            newpos = (LONGLONG)lpParms->dwTo * 10000;
+        }
     } else {
         WARN("dwFlag doesn't tell where to seek to...\n");
         return MCIERR_MISSING_PARAMETER;
@@ -788,25 +804,52 @@ static DWORD MCIQTZ_mciSet(UINT wDevID, DWORD dwFlags, LPMCI_DGV_SET_PARMS lpPar
         }
     }
 
-    if (dwFlags & MCI_SET_DOOR_OPEN)
-        FIXME("MCI_SET_DOOR_OPEN not implemented yet\n");
-    if (dwFlags & MCI_SET_DOOR_CLOSED)
-        FIXME("MCI_SET_DOOR_CLOSED not implemented yet\n");
-    if (dwFlags & MCI_SET_AUDIO)
-        FIXME("MCI_SET_AUDIO not implemented yet\n");
+    /* a file has no door */
+    if (dwFlags & (MCI_SET_DOOR_OPEN | MCI_SET_DOOR_CLOSED))
+        return MCIERR_UNSUPPORTED_FUNCTION;
+
+    if ((dwFlags & (MCI_SET_AUDIO | MCI_SET_VIDEO)) && !(dwFlags & (MCI_SET_ON | MCI_SET_OFF)))
+        return MCIERR_MISSING_PARAMETER;
+
+    if (dwFlags & MCI_SET_AUDIO) {
+        BOOL off = !!(dwFlags & MCI_SET_OFF), was_off = wma->audio_off[0] && wma->audio_off[1];
+
+        switch (lpParms->dwAudio) {
+            case MCI_SET_AUDIO_ALL: wma->audio_off[0] = wma->audio_off[1] = off; break;
+            case MCI_SET_AUDIO_LEFT: wma->audio_off[0] = off; break;
+            case MCI_SET_AUDIO_RIGHT: wma->audio_off[1] = off; break;
+            default: return MCIERR_BAD_INTEGER;
+        }
+
+        /* both channels off is silence; one off moves the balance to the other side */
+        if (!was_off && wma->audio_off[0] && wma->audio_off[1]) {
+            IBasicAudio_get_Volume(wma->audio, &wma->volume);
+            IBasicAudio_put_Volume(wma->audio, -10000);
+        } else {
+            if (was_off && !(wma->audio_off[0] && wma->audio_off[1]))
+                IBasicAudio_put_Volume(wma->audio, wma->volume);
+            if (wma->audio_off[0] && !wma->audio_off[1])
+                IBasicAudio_put_Balance(wma->audio, 10000);
+            else if (wma->audio_off[1] && !wma->audio_off[0])
+                IBasicAudio_put_Balance(wma->audio, -10000);
+            else if (!wma->audio_off[0] && !wma->audio_off[1])
+                IBasicAudio_put_Balance(wma->audio, 0);
+        }
+    }
+
     if (dwFlags & MCI_SET_VIDEO)
-        FIXME("MCI_SET_VIDEO not implemented yet\n");
-    if (dwFlags & MCI_SET_ON)
-        FIXME("MCI_SET_ON not implemented yet\n");
-    if (dwFlags & MCI_SET_OFF)
-        FIXME("MCI_SET_OFF not implemented yet\n");
-    if (dwFlags & MCI_SET_AUDIO_LEFT)
-        FIXME("MCI_SET_AUDIO_LEFT not implemented yet\n");
-    if (dwFlags & MCI_SET_AUDIO_RIGHT)
-        FIXME("MCI_SET_AUDIO_RIGHT not implemented yet\n");
-
-    if (dwFlags & ~0x7f03 /* All MCI_SET flags mask */)
-        ERR("Unknown flags %08lx\n", dwFlags & ~0x7f03);
+        IVideoWindow_put_Visible(wma->vidwin, (dwFlags & MCI_SET_ON) ? OATRUE : OAFALSE);
+
+    /* the playback speed, in thousandths of the normal speed */
+    if (dwFlags & MCI_DGV_SET_SPEED) {
+        if (!lpParms->dwSpeed)
+            return MCIERR_OUTOFRANGE;
+        if (FAILED(IMediaSeeking_SetRate(wma->seek, lpParms->dwSpeed / 1000.0)))
+            return MCIERR_OUTOFRANGE;
+    }
+
+    if (dwFlags & ~(0x7f03 /* All MCI_SET flags mask */ | MCI_DGV_SET_SPEED))
+        WARN("Unknown flags %08lx\n", dwFlags & ~(0x7f03 | MCI_DGV_SET_SPEED));
 
     return 0;
 }
@@ -871,8 +914,9 @@ static DWORD MCIQTZ_mciStatus(UINT wDevID, DWORD dwFlags, LPMCI_DGV_STATUS_PARMS
             break;
         }
         case MCI_STATUS_NUMBER_OF_TRACKS:
-            FIXME("MCI_STATUS_NUMBER_OF_TRACKS not implemented yet\n");
-            return MCIERR_UNRECOGNIZED_COMMAND;
+            /* a file is a single track */
+            lpParms->dwReturn = 1;
+            break;
         case MCI_STATUS_MODE: {
             LONG state = State_Stopped;
             IMediaControl_GetState(wma->pmctrl, -1, &state);
@@ -888,19 +932,22 @@ static DWORD MCIQTZ_mciStatus(UINT wDevID, DWORD dwFlags, LPMCI_DGV_STATUS_PARMS
             break;
         }
         case MCI_STATUS_MEDIA_PRESENT:
-            FIXME("MCI_STATUS_MEDIA_PRESENT not implemented yet\n");
-            return MCIERR_UNRECOGNIZED_COMMAND;
+            lpParms->dwReturn = MAKEMCIRESOURCE(TRUE, MCI_TRUE);
+            ret = MCI_RESOURCE_RETURNED;
+            break;
         case MCI_STATUS_TIME_FORMAT:
             lpParms->dwReturn = MAKEMCIRESOURCE(wma->time_format,
                                                 MCI_FORMAT_RETURN_BASE + wma->time_format);
             ret = MCI_RESOURCE_RETURNED;
             break;
         case MCI_STATUS_READY:
-            FIXME("MCI_STATUS_READY not implemented yet\n");
-            return MCIERR_UNRECOGNIZED_COMMAND;
+            /* an open file is ready */
+            lpParms->dwReturn = MAKEMCIRESOURCE(TRUE, MCI_TRUE);
+            ret = MCI_RESOURCE_RETURNED;
+            break;
         case MCI_STATUS_CURRENT_TRACK:
-            FIXME("MCI_STATUS_CURRENT_TRACK not implemented yet\n");
-            return MCIERR_UNRECOGNIZED_COMMAND;
+            lpParms->dwReturn = 1;
+            break;
         default:
             FIXME("Unknown command %08lX\n", lpParms->dwItem);
             return MCIERR_UNRECOGNIZED_COMMAND;
@@ -939,29 +986,30 @@ static DWORD MCIQTZ_mciWhere(UINT wDevID, DWORD dwFlags, LPMCI_DGV_RECT_PARMS lp
     }
 
     if (dwFlags & MCI_DGV_WHERE_SOURCE) {
-        if (dwFlags & MCI_DGV_WHERE_MAX)
-            FIXME("MCI_DGV_WHERE_SOURCE_MAX stub\n");
-        IBasicVideo_GetSourcePosition(wma->vidbasic, &rc.left, &rc.top, &rc.right, &rc.bottom);
-        TRACE("MCI_DGV_WHERE_SOURCE %s\n", wine_dbgstr_rect(&rc));
+        if (dwFlags & MCI_DGV_WHERE_MAX) {
+            /* the whole frame of the video */
+            rc.left = rc.top = 0;
+            IBasicVideo_GetVideoSize(wma->vidbasic, &rc.right, &rc.bottom);
+            TRACE("MCI_DGV_WHERE_SOURCE_MAX %s\n", wine_dbgstr_rect(&rc));
+        } else {
+            IBasicVideo_GetSourcePosition(wma->vidbasic, &rc.left, &rc.top, &rc.right, &rc.bottom);
+            TRACE("MCI_DGV_WHERE_SOURCE %s\n", wine_dbgstr_rect(&rc));
+        }
     }
     if (dwFlags & MCI_DGV_WHERE_DESTINATION) {
-        if (dwFlags & MCI_DGV_WHERE_MAX)
-            FIXME("MCI_DGV_WHERE_DESTINATION_MAX stub\n");
-        IBasicVideo_GetDestinationPosition(wma->vidbasic, &rc.left, &rc.top, &rc.right, &rc.bottom);
-        TRACE("MCI_DGV_WHERE_DESTINATION %s\n", wine_dbgstr_rect(&rc));
-    }
-    if (dwFlags & MCI_DGV_WHERE_FRAME) {
-        if (dwFlags & MCI_DGV_WHERE_MAX)
-            FIXME("MCI_DGV_WHERE_FRAME_MAX not supported yet\n");
-        else
-            FIXME("MCI_DGV_WHERE_FRAME not supported yet\n");
-        goto out;
+        if (dwFlags & MCI_DGV_WHERE_MAX) {
+            /* all of the window's client area */
+            GetClientRect(hWnd ? hWnd : wma->parent, &rc);
+            TRACE("MCI_DGV_WHERE_DESTINATION_MAX %s\n", wine_dbgstr_rect(&rc));
+        } else {
+            IBasicVideo_GetDestinationPosition(wma->vidbasic, &rc.left, &rc.top, &rc.right, &rc.bottom);
+            TRACE("MCI_DGV_WHERE_DESTINATION %s\n", wine_dbgstr_rect(&rc));
+        }
     }
-    if (dwFlags & MCI_DGV_WHERE_VIDEO) {
-        if (dwFlags & MCI_DGV_WHERE_MAX)
-            FIXME("MCI_DGV_WHERE_VIDEO_MAX not supported yet\n");
-        else
-            FIXME("MCI_DGV_WHERE_VIDEO not supported yet\n");
+    if (dwFlags & (MCI_DGV_WHERE_FRAME | MCI_DGV_WHERE_VIDEO)) {
+        /* frame and video buffers belong to capture hardware */
+        WARN("MCI_DGV_WHERE_FRAME / MCI_DGV_WHERE_VIDEO are not supported by file playback\n");
+        ret = MCIERR_UNSUPPORTED_FUNCTION;
         goto out;
     }
     if (dwFlags & MCI_DGV_WHERE_WINDOW) {
@@ -1210,6 +1258,155 @@ static DWORD MCIQTZ_mciSetAudio(UINT wDevID, DWORD dwFlags, LPMCI_DGV_SETAUDIO_P
     return ret;
 }
 
+/***************************************************************************
+ *                              MCIQTZ_mciStep                  [internal]
+ */
+static DWORD MCIQTZ_mciStep(UINT wDevID, DWORD dwFlags, LPMCI_DGV_STEP_PARMS lpParms)
+{
+    WINE_MCIQTZ *wma = MCIQTZ_mciGetOpenDev(wDevID);
+    LONGLONG pos, duration;
+    REFTIME frame = 0.0;
+    LONG frames = 1;
+
+    TRACE("(%04x, %08lX, %p)\n", wDevID, dwFlags, lpParms);
+
+    if (!lpParms)
+        return MCIERR_NULL_PARAMETER_BLOCK;
+    if (!wma)
+        return MCIERR_INVALID_DEVICE_ID;
+    if (dwFlags & MCI_TEST)
+        return 0;
+
+    if (FAILED(IBasicVideo_get_AvgTimePerFrame(wma->vidbasic, &frame)) || frame <= 0.0)
+        return MCIERR_UNSUPPORTED_FUNCTION;
+    if (dwFlags & MCI_DGV_STEP_FRAMES)
+        frames = lpParms->dwFrames;
+    if (dwFlags & MCI_DGV_STEP_REVERSE)
+        frames = -frames;
+
+    /* stepping leaves the video paused on the new frame */
+    IMediaControl_Pause(wma->pmctrl);
+    IMediaSeeking_SetTimeFormat(wma->seek, &TIME_FORMAT_MEDIA_TIME);
+    if (FAILED(IMediaSeeking_GetCurrentPosition(wma->seek, &pos)))
+        return MCIERR_INTERNAL;
+    IMediaSeeking_GetDuration(wma->seek, &duration);
+    pos += (LONGLONG)(frames * frame * 10000000.0);
+    pos = max(0, min(pos, duration));
+    if (FAILED(IMediaSeeking_SetPositions(wma->seek, &pos, AM_SEEKING_AbsolutePositioning, NULL, AM_SEEKING_NoPositioning)))
+        return MCIERR_INTERNAL;
+
+    if (dwFlags & MCI_NOTIFY)
+        MCIQTZ_mciNotify(lpParms->dwCallback, wma, MCI_NOTIFY_SUCCESSFUL);
+    return 0;
+}
+
+/***************************************************************************
+ *                              MCIQTZ_mciCue                   [internal]
+ */
+static DWORD MCIQTZ_mciCue(UINT wDevID, DWORD dwFlags, LPMCI_DGV_CUE_PARMS lpParms)
+{
+    WINE_MCIQTZ *wma = MCIQTZ_mciGetOpenDev(wDevID);
+    LONGLONG pos;
+    REFTIME frame;
+
+    TRACE("(%04x, %08lX, %p)\n", wDevID, dwFlags, lpParms);
+
+    if (!lpParms)
+        return MCIERR_NULL_PARAMETER_BLOCK;
+    if (!wma)
+        return MCIERR_INVALID_DEVICE_ID;
+    /* there is nothing to record from */
+    if (dwFlags & MCI_DGV_CUE_INPUT)
+        return MCIERR_UNSUPPORTED_FUNCTION;
+    if (dwFlags & MCI_TEST)
+        return 0;
+
+    /* cueing for output: paused and ready to play, at the position asked for */
+    if (dwFlags & MCI_TO)
+    {
+        IMediaSeeking_SetTimeFormat(wma->seek, &TIME_FORMAT_MEDIA_TIME);
+        if (wma->time_format == MCI_FORMAT_FRAMES)
+        {
+            if (FAILED(IBasicVideo_get_AvgTimePerFrame(wma->vidbasic, &frame)) || frame <= 0.0)
+                return MCIERR_UNSUPPORTED_FUNCTION;
+            pos = (LONGLONG)(lpParms->dwTo * frame * 10000000.0 + 0.5);
+        }
+        else
+            pos = (LONGLONG)lpParms->dwTo * 10000;
+        IMediaSeeking_SetPositions(wma->seek, &pos, AM_SEEKING_AbsolutePositioning, NULL, AM_SEEKING_NoPositioning);
+    }
+    IMediaControl_Pause(wma->pmctrl);
+
+    if (dwFlags & MCI_NOTIFY)
+        MCIQTZ_mciNotify(lpParms->dwCallback, wma, MCI_NOTIFY_SUCCESSFUL);
+    return 0;
+}
+
+/***************************************************************************
+ *                              MCIQTZ_mciInfo                  [internal]
+ */
+static DWORD MCIQTZ_mciInfo(UINT wDevID, DWORD dwFlags, LPMCI_DGV_INFO_PARMSW lpParms)
+{
+    WINE_MCIQTZ *wma = MCIQTZ_mciGetOpenDev(wDevID);
+    WCHAR caption[256];
+    const WCHAR *str;
+
+    TRACE("(%04x, %08lX, %p)\n", wDevID, dwFlags, lpParms);
+
+    if (!lpParms || !lpParms->lpstrReturn)
+        return MCIERR_NULL_PARAMETER_BLOCK;
+    if (!wma)
+        return MCIERR_INVALID_DEVICE_ID;
+
+    if (dwFlags & MCI_INFO_PRODUCT)
+        str = L"DirectShow";
+    else if (dwFlags & MCI_INFO_FILE)
+        str = wma->file ? wma->file : L"";
+    else if (dwFlags & MCI_DGV_INFO_TEXT)
+    {
+        GetWindowTextW(wma->parent, caption, ARRAY_SIZE(caption));
+        str = caption;
+    }
+    else
+        return MCIERR_UNSUPPORTED_FUNCTION;
+
+    if (!lpParms->dwRetSize)
+        return MCIERR_PARAM_OVERFLOW;
+    lstrcpynW(lpParms->lpstrReturn, str, lpParms->dwRetSize);
+    if (wcslen(str) >= lpParms->dwRetSize)
+        return MCIERR_PARAM_OVERFLOW;
+
+    if (dwFlags & MCI_NOTIFY)
+        MCIQTZ_mciNotify(lpParms->dwCallback, wma, MCI_NOTIFY_SUCCESSFUL);
+    return 0;
+}
+
+/***************************************************************************
+ *                              MCIQTZ_mciSetVideo              [internal]
+ */
+static DWORD MCIQTZ_mciSetVideo(UINT wDevID, DWORD dwFlags, LPMCI_DGV_SETVIDEO_PARMSW lpParms)
+{
+    WINE_MCIQTZ *wma = MCIQTZ_mciGetOpenDev(wDevID);
+
+    TRACE("(%04x, %08lX, %p)\n", wDevID, dwFlags, lpParms);
+
+    if (!lpParms)
+        return MCIERR_NULL_PARAMETER_BLOCK;
+    if (!wma)
+        return MCIERR_INVALID_DEVICE_ID;
+
+    if (dwFlags & (MCI_SET_ON | MCI_SET_OFF))
+        IVideoWindow_put_Visible(wma->vidwin, (dwFlags & MCI_SET_ON) ? OATRUE : OAFALSE);
+
+    /* picture adjustments (brightness, colour, ...) need display hardware */
+    if (dwFlags & MCI_DGV_SETVIDEO_ITEM)
+        return MCIERR_UNSUPPORTED_FUNCTION;
+
+    if (dwFlags & MCI_NOTIFY)
+        MCIQTZ_mciNotify(lpParms->dwCallback, wma, MCI_NOTIFY_SUCCESSFUL);
+    return 0;
+}
+
 /*======================================================================*
  *                          MCI QTZ entry points                        *
  *======================================================================*/
@@ -1260,32 +1457,39 @@ LRESULT CALLBACK MCIQTZ_DriverProc(DWORD_PTR dwDevID, HDRVR hDriv, UINT wMsg,
             return MCIQTZ_mciWindow(dwDevID, dwParam1, (LPMCI_DGV_WINDOW_PARMSW)dwParam2);
         case MCI_PUT:
             return MCIQTZ_mciPut(dwDevID, dwParam1, (MCI_GENERIC_PARMS*)dwParam2);
-        case MCI_RECORD:
+        case MCI_STEP:
+            return MCIQTZ_mciStep(dwDevID, dwParam1, (LPMCI_DGV_STEP_PARMS)dwParam2);
+        case MCI_CUE:
+            return MCIQTZ_mciCue(dwDevID, dwParam1, (LPMCI_DGV_CUE_PARMS)dwParam2);
         case MCI_INFO:
+            return MCIQTZ_mciInfo(dwDevID, dwParam1, (LPMCI_DGV_INFO_PARMSW)dwParam2);
+        case MCI_SETVIDEO:
+            return MCIQTZ_mciSetVideo(dwDevID, dwParam1, (LPMCI_DGV_SETVIDEO_PARMSW)dwParam2);
+        case MCI_REALIZE:
+            /* the video renderer handles its own palette */
+            return MCIQTZ_mciGetOpenDev(dwDevID) ? 0 : MCIERR_INVALID_DEVICE_ID;
+        /* recording, editing and capture hardware: not a file player's job */
+        case MCI_RECORD:
         case MCI_LOAD:
         case MCI_SAVE:
         case MCI_FREEZE:
-        case MCI_REALIZE:
         case MCI_UNFREEZE:
-        case MCI_STEP:
         case MCI_COPY:
         case MCI_CUT:
         case MCI_DELETE:
         case MCI_PASTE:
-        case MCI_CUE:
         /* Digital Video specific */
         case MCI_CAPTURE:
         case MCI_MONITOR:
         case MCI_RESERVE:
         case MCI_SIGNAL:
-        case MCI_SETVIDEO:
         case MCI_QUALITY:
         case MCI_LIST:
         case MCI_UNDO:
         case MCI_CONFIGURE:
         case MCI_RESTORE:
-            FIXME("Unimplemented command [%08X]\n", wMsg);
-            break;
+            WARN("Unsupported command [%08X]\n", wMsg);
+            return MCIERR_UNSUPPORTED_FUNCTION;
         case MCI_SPIN:
         case MCI_ESCAPE:
             WARN("Unsupported command [%08X]\n", wMsg);
diff --git a/dlls/mciqtz32/mciqtz_private.h b/dlls/mciqtz32/mciqtz_private.h
index 9206a88..01b87e6 100644
--- a/dlls/mciqtz32/mciqtz_private.h
+++ b/dlls/mciqtz32/mciqtz_private.h
@@ -47,6 +47,10 @@ typedef struct {
     HANDLE         callback;
     HANDLE         thread;
     HANDLE         stop_event;
+    WCHAR         *file;          /* for MCI_INFO */
+    BOOL           audio_off[2];  /* left, right turned off with MCI_SET_AUDIO */
+    LONG           volume;        /* volume while the audio is turned off */
+    LONG           balance;
 } WINE_MCIQTZ;
 
 #endif  /* __WINE_PRIVATE_MCIQTZ_H */
